File: bézier/julia/05-bezier-matrix-sympy-text.jl
# ============================================================
# Bezier Curve — Matrix Method  [P(u) = U · M · G]
# Julia port of 05-bezier-matrix-sympy-text.py
# Exact rational arithmetic via built-in Rational{Int}
# Polynomial formatting via Symbolics.jl
# Table formatting via plain strings (no external deps for tables)
# ============================================================

using Symbolics
using Printf

@variables u


# ============================================================
# -- Switch  (only change this one line) ---------------------
# ============================================================

const CASE = 2   # 1 = Quadratic 2D,  2 = Quartic 3D


# ============================================================
# -- Case Data  (do not change) ------------------------------
# ============================================================

const CASE_DATA = Dict(
    1 => (
        points   = Rational{Int}[1 1;
                                  3 3;
                                  5 1],
        u_values = Rational{Int}[0, 1//4, 1//2, 3//4, 1],
        axes     = ["x", "y"],
    ),
    2 => (
        points   = Rational{Int}[0 0 1;
                                  0 4 1;
                                  4 0 1;
                                  4 4 1;
                                  5 4 1],
        u_values = Rational{Int}[0, 1//4, 1//2, 3//4, 1],
        axes     = ["x", "y", "z"],
    ),
)


# ============================================================
# -- BezierMatrix --------------------------------------------
# ============================================================

struct BezierMatrix
    G      ::Matrix{Rational{Int}}
    axes   ::Vector{String}
    n      ::Int
    degree ::Int
    dim    ::Int
    M      ::Matrix{Rational{Int}}
    MG     ::Matrix{Rational{Int}}
end

"""
    BezierMatrix(points, axes)

Exact rational Bezier matrix.  M derived from Bernstein via:
  M[n-k, i] = C(n,i) · C(n-i, k-i) · (-1)^(k-i)   for k >= i
All matrix products stay in Rational{Int} — no floats.
"""
function BezierMatrix(points::Matrix{Rational{Int}}, axes::Vector{String})
    n   = size(points, 1)
    deg = n - 1
    M   = derive_M(deg)
    BezierMatrix(points, axes, n, deg, length(axes), M, M * points)
end

function derive_M(deg::Int)::Matrix{Rational{Int}}
    n = deg
    M = zeros(Rational{Int}, n+1, n+1)
    for i in 0:n
        m = n - i
        for j in 0:m
            k = i + j
            M[n-k+1, i+1] = binomial(n,i) * binomial(m,j) * (-1)^j
        end
    end
    return M
end

function evaluate_all(bm::BezierMatrix, u_values::Vector{Rational{Int}})::Matrix{Rational{Int}}
    n    = bm.degree
    Uall = [u_val^(n-k) for u_val in u_values, k in 0:n]
    Uall * bm.MG
end


# ============================================================
# -- Formatting helpers --------------------------------------
# ============================================================

const WIDTH = 66

ruler(char::Char='=') = string(char)^WIDTH

# Format a Rational — integer if denominator is 1, else p//q notation
fmt_rat(r::Rational{Int}) = isone(denominator(r)) ? string(numerator(r)) : string(r)
fmt_rat(n::Integer)       = string(n)

section(title::String) = (
    println();
    println("  $title");
    println("  " * "-"^(WIDTH-2))
)

# ── Polynomial formatting via Symbolics.jl ───────────────────────────────────
# MG column k holds coefficient of u^(degree-k), highest power first.
# We build a symbolic expression by summing coeff * u^power, then expand().

function fmt_poly(mg_col::Vector{Rational{Int}}, deg::Int)::String
    expr = sum(
        mg_col[k+1] * u^(deg-k)
        for k in 0:deg
        if mg_col[k+1] != 0
    ; init = 0*u^0)
    string(expand(expr))
end

# ── Plain-text table printer ─────────────────────────────────────────────────
# style :simple  → header + underline
# style :outline → box border

function print_table(
        rows       ::Vector{Vector{String}},
        headers    ::Vector{String};
        row_labels ::Union{Vector{String}, Nothing} = nothing,
        style      ::Symbol = :simple)

    has_rl = row_labels !== nothing
    widths = [length(h) for h in headers]
    for row in rows
        for (c, cell) in enumerate(row)
            widths[c] = max(widths[c], length(cell))
        end
    end
    rl_w = has_rl ? max(maximum(length.(row_labels)), 2) : 0

    function fmt_row(cells, label=nothing)
        parts = String[]
        has_rl && push!(parts, lpad(something(label, ""), rl_w))
        for (c, cell) in enumerate(cells)
            push!(parts, lpad(cell, widths[c]))
        end
        "  " * join(parts, "  ")
    end

    hdr     = fmt_row(headers, has_rl ? "" : nothing)
    ubar    = "  " * (has_rl ? " "^rl_w * "  " : "") * join(["-"^w for w in widths], "  ")
    inner_w = sum(widths) + 2*(length(headers)-1) + (has_rl ? rl_w + 2 : 0)

    if style == :outline
        println("  ╒" * "═"^(inner_w+2) * "╕")
        println("  │ " * lstrip(hdr) * " │")
        println("  ╞" * "═"^(inner_w+2) * "╡")
        for (r, row) in enumerate(rows)
            println("  │ " * lstrip(fmt_row(row, has_rl ? row_labels[r] : nothing)) * " │")
        end
        println("  ╘" * "═"^(inner_w+2) * "╛")
    else
        println(hdr)
        println(ubar)
        for (r, row) in enumerate(rows)
            println(fmt_row(row, has_rl ? row_labels[r] : nothing))
        end
    end
end

function print_labeled_matrix(
        mat        ::Matrix{Rational{Int}},
        row_labels ::Vector{String},
        col_labels ::Vector{String})
    rows = [[fmt_rat(mat[r,c]) for c in 1:size(mat,2)] for r in 1:size(mat,1)]
    print_table(rows, col_labels; row_labels=row_labels)
end


# ============================================================
# -- Printer functions ---------------------------------------
# ============================================================

function print_title(bm::BezierMatrix, case_num::Int)
    names = Dict(2=>"Quadratic", 3=>"Cubic", 4=>"Quartic", 5=>"Quintic")
    dname = get(names, bm.degree, "Degree-$(bm.degree)")
    println()
    println(ruler())
    println("  Bezier Curve — Matrix Method  [P(u) = U · M · G]  (Symbolics.jl)")
    println("  Case $case_num  |  $dname  |  $(bm.dim)D  |  $(bm.n) Control Points")
    println(ruler())
end

function print_G(bm::BezierMatrix)
    section("G  — Geometry Matrix (Control Points)")
    println()
    print_labeled_matrix(bm.G, ["P$(i-1)" for i in 1:bm.n], bm.axes)
end

function print_M(bm::BezierMatrix)
    section("M  — Basis Matrix (derived via binomial)")
    n = bm.degree
    println()
    print_labeled_matrix(bm.M, ["u^$(n-k)" for k in 0:n], ["P$(i-1)" for i in 1:bm.n])
end

function print_MG(bm::BezierMatrix)
    section("M · G  — Polynomial Coefficient Matrix")
    n = bm.degree
    println()
    print_labeled_matrix(bm.MG, ["u^$(n-k)" for k in 0:n], bm.axes)
end

function print_polynomials(bm::BezierMatrix)
    section("Resulting Polynomials  (via Symbolics.jl)")
    println()
    for (d, axis) in enumerate(bm.axes)
        println("  $(axis)(u)  =  $(fmt_poly(bm.MG[:,d], bm.degree))")
    end
end

function print_eval_table(bm::BezierMatrix, u_values::Vector{Rational{Int}})
    section("Evaluation Table  —  P(u) = U · (M · G)")
    pts  = evaluate_all(bm, u_values)
    rows = Vector{Vector{String}}()
    for (r, u_val) in enumerate(u_values)
        row = [fmt_rat(u_val)]
        for c in 1:bm.dim
            exact = pts[r, c]
            dec   = @sprintf("%.4f", Float64(exact))
            push!(row, "$(fmt_rat(exact))  ($dec)")
        end
        push!(rows, row)
    end
    println()
    print_table(rows, vcat(["u"], ["$(ax)(u)" for ax in bm.axes]); style=:outline)
end

function print_all(bm::BezierMatrix, u_values::Vector{Rational{Int}}, case_num::Int)
    print_title(bm, case_num)
    print_G(bm)
    print_M(bm)
    print_MG(bm)
    print_polynomials(bm)
    print_eval_table(bm, u_values)
    println()
    println(ruler())
    println()
end


# ============================================================
# -- Entry Point ---------------------------------------------
# ============================================================

function main()
    data = CASE_DATA[CASE]
    bm   = BezierMatrix(data.points, data.axes)
    print_all(bm, data.u_values, CASE)
end

main()

File: bézier/python/05-bezier-matrix-sympy-text.py
import sympy as sp
from tabulate import tabulate


# ============================================================
# -- Switch  (only change this one line) ---------------------
# ============================================================

CASE = 2   # 1 = Quadratic 2D,  2 = Quartic 3D


# ============================================================
# -- Case Data  (do not change) ------------------------------
# ============================================================

CASE_DATA = {
    1: dict(
        points   = [
            [1, 1],   # P0
            [3, 3],   # P1
            [5, 1],   # P2
        ],
        u_values = [0, sp.Rational(1,4), sp.Rational(1,2),
                       sp.Rational(3,4), 1],
        axes     = ['x', 'y'],
    ),
    2: dict(
        points   = [
            [0, 0, 1],   # P0
            [0, 4, 1],   # P1
            [4, 0, 1],   # P2
            [4, 4, 1],   # P3
            [5, 4, 1],   # P4
        ],
        u_values = [0, sp.Rational(1,4), sp.Rational(1,2),
                       sp.Rational(3,4), 1],
        axes     = ['x', 'y', 'z'],
    ),
}


# ============================================================
# -- Bezier Matrix  (generic — do not change) ----------------
# ============================================================

class BezierMatrix:
    """
    Evaluates a Bezier curve of any degree via P(u) = U · M · G
    using SymPy matrices throughout — exact rational arithmetic,
    no floating-point error.

    SymPy replaces numpy here:

      _derive_M()
        numpy zeros + manual comb()  →  sp.Matrix built directly
        from sp.binomial(), with sp.Rational entries.
        All coefficients stay as exact integers.

      G
        numpy array of floats  →  sp.Matrix of sp.Rational,
        so matrix products stay exact end-to-end.

      MG  =  M · G
        numpy @  →  sp.Matrix.__mul__  (exact rational result)

      evaluate_all()
        numpy Vandermonde loop  →  sp.Matrix of u^k powers
        using sp.Rational u_values, giving exact results.

    M[n-k, i]  =  C(n,i) · C(n-i, k-i) · (-1)^(k-i)
                  for k >= i,  else 0
    Row 0 = highest power u^n,  last row = constant u^0.
    """

    u = sp.Symbol('u')   # shared SymPy symbol

    def __init__(self, points: list, axes: list):
        self.axes   = axes
        self.n      = len(points)
        self.degree = self.n - 1
        self.dim    = len(axes)

        # G — geometry matrix, exact rational entries
        self.G  = sp.Matrix([
            [sp.Rational(c) for c in row]
            for row in points
        ])

        # M — basis matrix derived from Bernstein
        self.M  = self._derive_M()

        # M·G — polynomial coefficient matrix, computed once
        self.MG = self.M * self.G

    def _derive_M(self) -> sp.Matrix:
        """
        Derive the Bezier basis matrix M of shape (n+1, n+1).

        SymPy's sp.binomial() computes C(n,i) exactly as an
        integer — no manual comb() needed.  The result is a
        sp.Matrix with exact integer entries.
        """
        n    = self.degree
        size = n + 1
        M    = sp.zeros(size, size)

        for i in range(size):
            m = n - i
            for j in range(m + 1):
                k        = i + j
                M[n-k, i] = (sp.binomial(n, i)
                              * sp.binomial(m, j)
                              * (-1)**j)
        return M

    def U(self, u_val) -> sp.Matrix:
        """
        Parameter row vector  [u^n,  u^(n-1),  ...,  u,  1]
        as a (1 × n+1) SymPy matrix.
        u_val may be a sp.Rational or plain integer.
        """
        n = self.degree
        return sp.Matrix([[u_val ** (n - k)
                           for k in range(n + 1)]])

    def evaluate_all(self, u_values: list) -> sp.Matrix:
        """
        Evaluate at all u values in one batched matrix multiply.

        Builds U_all of shape (m, n+1) where each row is the
        power vector for one u value, then:
          P  =  U_all · M · G  =  U_all · MG

        All arithmetic stays in SymPy — results are exact
        rationals, not floats.
        """
        U_all = sp.Matrix([
            [u_val ** (self.degree - k)
             for k in range(self.degree + 1)]
            for u_val in u_values
        ])
        return U_all * self.MG   # (m, n+1) · (n+1, dim) → (m, dim)


# ============================================================
# -- Console Printer  (generic — do not change) --------------
# ============================================================

class MatrixPrinter:
    """
    Prints G, M, M·G, polynomials, and evaluation table
    using tabulate for all layout and SymPy for all math.

    What SymPy provides vs the numpy+tabulate version:
      _derive_M      → sp.Matrix with sp.binomial (exact int)
      MG             → sp.Matrix product  (exact rational)
      evaluate_all   → sp.Matrix product  (exact rational)
      _fmt_poly      → sp.Poly.as_expr()  (auto-formatted)
      cell values    → str(sp.Rational)   (exact fractions)
    """

    WIDTH = 66

    def __init__(
        self,
        bm:       BezierMatrix,
        u_values: list,
        case_num: int,
    ):
        self.bm       = bm
        self.u_values = u_values
        self.case_num = case_num

    # -- Formatting helpers ----------------------------------

    def _ruler(self, char: str = '=') -> str:
        return char * self.WIDTH

    def _section(self, title: str):
        print()
        print(f'  {title}')
        print('  ' + '-' * (self.WIDTH - 2))

    def _fmt(self, v) -> str:
        """
        Format a SymPy number.
        Integers print without denominator, rationals as p/q.
        str() on a SymPy Rational already does the right thing.
        """
        return str(v)

    def _fmt_poly(self, col: int) -> str:
        """
        Build the polynomial string for one axis column of MG.

        MG[:, col] gives the monomial coefficients ordered
        from highest power (row 0 = u^n) to constant (last row).

        sp.Poly needs them in the same order SymPy expects —
        highest power first — which is exactly what MG provides.

        sp.Poly(coeffs, u).as_expr() returns a clean SymPy
        expression; str() renders it in standard notation.
        SymPy handles sign, zero suppression, and coefficient-
        of-1 automatically — no manual formatting needed.
        """
        n      = self.bm.degree
        coeffs = [self.bm.MG[k, col] for k in range(n + 1)]
        expr   = sp.Poly(coeffs, self.bm.u).as_expr()
        return str(expr)

    def _mat_rows(self, mat: sp.Matrix, row_labels: list) -> list:
        """
        Convert a SymPy matrix to list-of-lists for tabulate,
        with row labels prepended as the first column.
        str() on each SymPy entry gives exact fractions.
        """
        return [
            [rl] + [self._fmt(mat[r, c])
                    for c in range(mat.shape[1])]
            for r, rl in enumerate(row_labels)
        ]

    def _print_mat(
        self,
        mat:        sp.Matrix,
        row_labels: list,
        col_labels: list,
    ):
        """Print a labelled SymPy matrix using tabulate."""
        rows = self._mat_rows(mat, row_labels)
        print(tabulate(
            rows,
            headers  = [''] + col_labels,
            tablefmt = 'simple',
            colalign = ('left',) + ('right',) * mat.shape[1],
        ))

    # -- Print -----------------------------------------------

    def print_all(self):
        self._print_title()
        self._print_G()
        self._print_M()
        self._print_MG()
        self._print_polynomials()
        self._print_table()
        print()
        print(self._ruler())
        print()

    def _print_title(self):
        names = {2:'Quadratic', 3:'Cubic',
                 4:'Quartic',   5:'Quintic'}
        dname = names.get(self.bm.degree, f'Degree-{self.bm.degree}')
        print()
        print(self._ruler())
        print(f'  Bezier Curve — Matrix Method  '
              f'[P(u) = U · M · G]  (SymPy + tabulate)')
        print(f'  Case {self.case_num}  |  {dname}  |  '
              f'{self.bm.dim}D  |  {self.bm.n} Control Points')
        print(self._ruler())

    def _print_G(self):
        self._section('G  — Geometry Matrix (Control Points)')
        print()
        self._print_mat(
            self.bm.G,
            row_labels = [f'P{i}' for i in range(self.bm.n)],
            col_labels = self.bm.axes,
        )

    def _print_M(self):
        self._section('M  — Basis Matrix (derived via sp.binomial)')
        n = self.bm.degree
        print()
        self._print_mat(
            self.bm.M,
            row_labels = [f'u^{n-k}' for k in range(self.bm.n)],
            col_labels = [f'P{i}'    for i in range(self.bm.n)],
        )

    def _print_MG(self):
        self._section('M · G  — Polynomial Coefficient Matrix')
        n = self.bm.degree
        print()
        self._print_mat(
            self.bm.MG,
            row_labels = [f'u^{n-k}' for k in range(self.bm.n)],
            col_labels = self.bm.axes,
        )

    def _print_polynomials(self):
        self._section('Resulting Polynomials  (via sp.Poly)')
        print()
        for d, axis in enumerate(self.bm.axes):
            print(f'  {axis}(u)  =  {self._fmt_poly(d)}')

    def _print_table(self):
        self._section('Evaluation Table  —  P(u) = U · (M · G)')

        # One batched SymPy matrix multiply — exact rationals
        pts  = self.bm.evaluate_all(self.u_values)
        rows = []
        for r, u_val in enumerate(self.u_values):
            row = [self._fmt(u_val)]
            for c in range(self.bm.dim):
                exact = pts[r, c]
                dec   = f'({float(exact):.4f})'
                row.append(f'{self._fmt(exact)}  {dec}')
            rows.append(row)

        headers = ['u'] + [f'{ax}(u)' for ax in self.bm.axes]
        print()
        print(tabulate(
            rows,
            headers  = headers,
            tablefmt = 'outline',
            colalign = ('right',) * (self.bm.dim + 1),
        ))


# ============================================================
# -- Entry Point ---------------------------------------------
# ============================================================

def main() -> int:
    data = CASE_DATA[CASE]

    bm = BezierMatrix(
        points = data['points'],
        axes   = data['axes'],
    )

    printer = MatrixPrinter(
        bm       = bm,
        u_values = data['u_values'],
        case_num = CASE,
    )

    printer.print_all()
    return 0


if __name__ == '__main__':
    raise SystemExit(main())
